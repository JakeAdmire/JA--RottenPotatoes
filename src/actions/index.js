export const setRestaurants = (restaurants) => ({
  type: 'SET_RESTAURANTS',
  restaurants
})

export const setLocation = (location) => ({
  type: 'SET_LOCATION',
  location
})

export const isLoading = (boolean) => ({
  type: 'IS_LOADING',
  boolean
})

export const setCard = (card) =>({
  type: 'SET_CARD',
  card
})