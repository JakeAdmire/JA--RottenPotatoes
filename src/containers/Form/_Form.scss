.Form {
  display: flex;
  flex-direction: column;
  width: 450px;

  h1 {
    font-family: 'Lobster', cursive;
    font-size: 80px;
    color: white;
    font-weight: 500;
    text-decoration: underline;
    text-shadow: 0 3px 6px rgba(0, 0, 0, 0.25);
    width: 580px;
    align-self: center;
    text-align: center;
    margin: 0 0 10px;
  }

  label {
    font-size: 22px;
    font-weight: bold;
    color: #404040;
    text-shadow: 0 3px 12px rgba(0, 0, 0, 0.25);
  }

  input, button {
    height: 38px;
    width: 100%;
    box-sizing: border-box;
    padding: 5px;
    border: none;
    color: #D9D9D9;
    background: #595959;
    font-size: 22px;
    box-shadow: 0 3px 12px rgba(0, 0, 0, 0.25);
    transition: all .2s ease;

    &:focus { outline: none; }
    &:hover { background: #404040; }
  }

  h3 {
    align-self: center;
    font-size: 31px;
    font-weight: bold;
    color: #404040;
    margin: 10px;
    text-shadow: 0 3px 12px rgba(0, 0, 0, .25);
  }

  button {
    margin: 0 0 50px;

    &:hover { cursor: pointer; }
  }
}