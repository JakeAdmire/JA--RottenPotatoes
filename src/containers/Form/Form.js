import React, { Component } from 'react';
import Geocode from "react-geocode";
import { connect } from 'react-redux';
import { setRestaurants, setLocation, setRedirect } from '../../actions';
import { geocodeKey } from '../../keys';

export class Form extends Component {
  constructor() {
    super();
    this.state = { value: '' };
  }

  componentDidMount() {
    document.addEventListener('keydown', this.handleEnter); 
  }

  handleEnter = (event) => {
    if (event.key !== 'Enter') return null;
    this.state.value && this.getLocation()
  }

  handleChange = (event) => {
    const { value } = event.target;
    this.setState({value});
  }

  getLocation = async () => {
    Geocode.setApiKey(geocodeKey);
    const { value } = this.state;
    try {
      const response = await Geocode.fromAddress(value);
      this.gatherLocationInfo(response.results[0])
    } catch (error) {
      console.log(error.message);
    }
  }

  gatherLocationInfo = (result) => {
    let location = {
      name: result.formatted_address,
      coords: result.geometry.location
    };
    this.props.setLocation(location);
    this.props.setRedirect('locations');
  }

  handleClick = (event) => {
    if ("geolocation" in navigator) {
      console.log('finding position...'); // start loader here
      navigator.geolocation.getCurrentPosition(
        function(position) { 
          const { latitude, longitude } = position.coords;
          callFunction({latitude, longitude});
        }
      );
      const callFunction = (position) => { this.getAddress(position); }
    } else {
      console.log('geolocation is not enabled on this browser');
    }
  }

  getAddress = async (position) => {
    Geocode.setApiKey(geocodeKey);
    try {
      const response = await Geocode.fromLatLng(position.latitude, position.longitude);
      this.gatherLocationInfo(response.results[0])
      console.log('...position found'); // end loader here
    } catch (error) {
      console.log(error.message);
    }
  }

  render() {
    const { value } = this.state;
    const placeholder='ex. 123 Fake Address, Denver, CO'
    return (
      <div className="Form">
        <label htmlFor="address">Enter an address</label>
        <input  value={value} 
                onChange={this.handleChange} 
                id="address" 
                type="text" 
                placeholder={placeholder}>
        </input>
        <h3>FOR</h3>
        <button onClick={this.handleClick}>Find My Location</button>
      </div>
    )
  }
}

export const setDispatchToProps = (dispatch) => ({
  setRestaurants: (restaurants) => dispatch(setRestaurants(restaurants)),
  setLocation: (location) => dispatch(setLocation(location)),
  setRedirect: (redirect) => dispatch(setRedirect(redirect))
})

export default connect(null, setDispatchToProps)(Form);