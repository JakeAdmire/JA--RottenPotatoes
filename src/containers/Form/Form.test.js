import React from 'react';
import { shallow } from 'enzyme';

import { Form, mapStateToProps, mapDispatchToProps } from './Form';
import { setRestaurants, setLocation, setRedirect, isLoading } from '../../actions';

describe('Form', () => {

  let wrapper;
  let mockSetLocation = jest.fn()

  beforeEach(() => {
    wrapper = shallow(
      <Form setLocation={mockSetLocation} />
    )
  })

  it('should match the snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  })

  it('should have default state', () => {
    let mockState = { value: '' };
    expect(wrapper.state()).toEqual(mockState);
  })

  it('should update state when handleChange is invoked', () => {
    const mockStateChange = { value: "test" }
    const mockEvent = { target: mockStateChange };
    wrapper.instance().handleChange(mockEvent);
    expect(wrapper.state()).toEqual(mockStateChange);
  })

  it.skip('should call handleClick on click', () => {
    const mockSpy = jest.spyOn(wrapper.instance(), 'handleClick');
    wrapper.update();

    wrapper.find('NavLink').simulate('click');
    expect(mockSpy).toBeCalled();
  })

  it.skip('should call handleEnter if key is pressed', () => {
    const mockEvent = { key: 'Enter' };
    const instance = wrapper.instance();
    let mockSpy = jest.spyOn(instance, 'handleEnter');
    wrapper.setState({ value: 'test'});
    wrapper.find('div').simulate('keypress', mockEvent);
    expect(mockSpy).toHaveBeenCalled();
  })

  it('should dispatch setLocation when gatherLocationInfo is called', () => {
    const mockResult = { 
      formatted_address: 'Denver',
      geometry: { location: {} } 
    };
    wrapper.instance().gatherLocationInfo(mockResult);
    expect(mockSetLocation).toHaveBeenCalled();
  })

  describe('mapStateToProps', () => {

    let mockState = { loading: false }
    
    it('should return a props object', () => {
      const results = mapStateToProps(mockState);
      expect(results).toEqual(mockState);
    })

  })

  describe('mapDispatchToProps', () => {

    let mockDispatch = jest.fn();

    it('should dispatch setRestaurants when its prop is called', () => {
      let mockRestaurants = [{}, {}];
      const actionToDispatch = setRestaurants(mockRestaurants);
      const mappedProps = mapDispatchToProps(mockDispatch);
      mappedProps.setRestaurants(mockRestaurants);
      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
    })

    it('should dispatch setLocation when its prop is called', () => {
      let mockLocation = { name: 'Hawaii', coords: {} };
      const actionToDispatch = setLocation(mockLocation);
      const mappedProps = mapDispatchToProps(mockDispatch);
      mappedProps.setLocation(mockLocation);
      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
    })

    it('should dispatch isLoading when its prop is called', () => {
      let mockBoolean = false;
      const actionToDispatch = isLoading(mockBoolean);
      const mappedProps = mapDispatchToProps(mockDispatch);
      mappedProps.isLoading(mockBoolean);
      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
    })

  })

})