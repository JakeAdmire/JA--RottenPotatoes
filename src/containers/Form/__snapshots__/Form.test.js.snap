// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form should match the snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Form
    isLoading={[MockFunction]}
    setLocation={[MockFunction]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <h1>
          Rotten Potatoes
        </h1>,
        <label
          htmlFor="address"
        >
          Enter an address:
        </label>,
        <input
          id="address"
          onChange={[Function]}
          placeholder="ex. 123 Fake Address, Denver, CO"
          type="text"
          value=""
        />,
        <h3>
          OR
        </h3>,
        <NavLink
          onClick={[Function]}
          to="/Rotten-Potatoes/locations"
        >
          Find My Location
        </NavLink>,
      ],
      "className": "Form",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Rotten Potatoes",
        },
        "ref": null,
        "rendered": "Rotten Potatoes",
        "type": "h1",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Enter an address:",
          "htmlFor": "address",
        },
        "ref": null,
        "rendered": "Enter an address:",
        "type": "label",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "id": "address",
          "onChange": [Function],
          "placeholder": "ex. 123 Fake Address, Denver, CO",
          "type": "text",
          "value": "",
        },
        "ref": null,
        "rendered": null,
        "type": "input",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "OR",
        },
        "ref": null,
        "rendered": "OR",
        "type": "h3",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": "Find My Location",
          "onClick": [Function],
          "to": "/Rotten-Potatoes/locations",
        },
        "ref": null,
        "rendered": "Find My Location",
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <h1>
            Rotten Potatoes
          </h1>,
          <label
            htmlFor="address"
          >
            Enter an address:
          </label>,
          <input
            id="address"
            onChange={[Function]}
            placeholder="ex. 123 Fake Address, Denver, CO"
            type="text"
            value=""
          />,
          <h3>
            OR
          </h3>,
          <NavLink
            onClick={[Function]}
            to="/Rotten-Potatoes/locations"
          >
            Find My Location
          </NavLink>,
        ],
        "className": "Form",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Rotten Potatoes",
          },
          "ref": null,
          "rendered": "Rotten Potatoes",
          "type": "h1",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Enter an address:",
            "htmlFor": "address",
          },
          "ref": null,
          "rendered": "Enter an address:",
          "type": "label",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "id": "address",
            "onChange": [Function],
            "placeholder": "ex. 123 Fake Address, Denver, CO",
            "type": "text",
            "value": "",
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "OR",
          },
          "ref": null,
          "rendered": "OR",
          "type": "h3",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": "Find My Location",
            "onClick": [Function],
            "to": "/Rotten-Potatoes/locations",
          },
          "ref": null,
          "rendered": "Find My Location",
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
